### PR TITLE
[Perf][GLM-5.3-Flash] Use FlashKDA for KDA chunked prefill (1.7-3.8x faster than the Triton chunk path)

### PR DESCRIPTION
**Pull request by JaredforReal**
## Purpose

One of three independent GLM-5.3-Flash perf PRs 

Use **FlashKDA** (`vllm._flashkda_C`, already built and used by Kimi-K3) for GLM-5.3-Flash's KDA chunked prefill instead of the ~15-kernel Triton `chunk_kda_with_fused_gate` path.

FlashKDA implements the same bounded-gate KDA recurrence GLM-5.3-Flash uses (`lower_bound * sigmoid(exp(A_log) * (g + dt_bias))`, in-kernel q/k l2norm, raw beta logits), so it is a drop-in. On GB300 per KDA layer (H=16/rank, D=128, `bench/flashkda_vs_chunk.py`):

| shape | Triton chunk path | FlashKDA |
|---|---|---|
| 8 x 2048 tokens | 183 us | 49 us |
| 4 x 8192 | 555 us | 146 us |
| 1 x 16384 | 1371 us | 787 us |

Selection: automatic when supported (SM90/SM10x/SM12x, bf16, head_dim 128, bounded gate); `additional_config.kda_prefill_backend = triton` keeps the old path (`flashkda` forces it). The output is written straight into the layer buffer (no merge copy) and the pre-sigmoid beta cast is not needed on this path. Spec-decode tokens themselves keep the Triton recurrent kernel (`num_accepted_tokens` rollback); in a step that carries both, FlashKDA runs the prefill segment into a workspace buffer and the outputs are scattered back by `non_spec_token_indx` / `spec_token_indx` (same scheme as the Kimi-K3 KDA layer; the merge writes straight into `core_attn_out`).

## Test Plan

- Unit / kernel tests listed under "Tests" below.
- Serving benchmark, all builds back-to-back in one session on 4x GB300 (SM100), TP4, prefix caching disabled:
  ```bash
  vllm serve zai-org/GLM-5.3-Flash -tp 4 --attention-backend FLASHINFER_MLA_SPARSE \
    --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384 \
    --gpu-memory-utilization 0.9 --no-enable-prefix-caching
  vllm bench serve --backend vllm --model glm53-flash --dataset-name random --random-range-ratio 0 \
    --ignore-eos --seed 0 --random-input-len <in> --random-output-len <out> --max-concurrency <c> \
    --num-prompts <n> --num-warmups <c> --percentile-metrics ttft,tpot,itl,e2el
  ```
  0 prefix-cache hit rate in the server log was checked to be 0.0% for every run.
- Accuracy per build: `lm_eval --model local-completions --tasks gsm8k --num_fewshot 5 --gen_kwargs temperature=0` (1319 questions), prompt-logprob agreement on real 4k/12k/30k/60k prompts vs main, and a 36-needle retrieval set at 6k/16k/30k (greedy, `reasoning_effort=low`), repeated 3x on main / +#55736 / full stack.

### Performance of this PR alone

Measured on top of #55736 (main + #55736) vs the same + this PR, 4x GB300, TP4, `--attention-backend FLASHINFER_MLA_SPARSE --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384`, **prefix caching disabled** (hit rate 0.0% checked in the server log), `vllm bench serve` random dataset with warmups, back-to-back in the same session. #55736 only touches decode, so this is the effect of FlashKDA on an otherwise-main prefill path. Decode tok/s is the steady-state window value; TPOT is the per-request median.

| point | metric | +#55736 | +#55736 +#55737 | delta |
|---|---|---|---|---|
| prefill 8x2048 | mean TTFT ms | 449 | 395 | -12.2% |
| prefill 2x8192 | mean TTFT ms | 380 | 330 | -13.2% |
| prefill 8x32768 | mean TTFT ms | 5108 | 4693 | -8.1% |
| prefill 2x65536 | mean TTFT ms | 2661 | 2450 | -7.9% |
| 32k/256 c=16 | decode tok/s | 296 | 316 | +6.6% |
| 32k/256 c=16 | mean TTFT ms | 3835 | 3642 | -5.0% |
| 32k/256 c=16 | median TPOT ms | 40.10 | 37.50 | -6.5% |
| 1k/512 c=1 | decode tok/s | 146 | 147 | +0.7% |
| 1k/512 c=1 | median TPOT ms | 6.46 | 6.47 | +0.2% |
| 1k/512 c=64 | decode tok/s | 3090 | 3132 | +1.4% |
| 1k/512 c=64 | median TPOT ms | 19.26 | 19.24 | -0.1% |
| 1k/512 c=256 | decode tok/s | 7367 | 7717 | +4.7% |
| 1k/512 c=256 | median TPOT ms | 32.87 | 31.30 | -4.8% |

### With speculative decoding (MTP k=1)

Same build, same settings plus `--speculative-config.method=mtp --speculative-config.num_speculative_tokens=1`, `kda_prefill_backend=triton` vs `flashkda`, back-to-back. Random-dataset acceptance is ~1.2 tokens/step, so absolute decode numbers are low; the comparison is what matters. Whole-run throughput (the steady-state window accounting is off under spec decode).

| point | metric | MTP k=1 + Triton KDA prefill  | MTP k=1 + FlashKDA |
|---|---|---|---|
| prefill 8x2048 | mean TTFT ms | 464 | 423 (-8.8%) |
| prefill 2x8192 | mean TTFT ms | 390 | 360 (-7.7%) |
| 32k/256 c=16 | decode tok/s | 260 | 284 (+9.5%) |
| 32k/256 c=16 | mean TTFT ms | 2623 | 2479 (-5.5%) |
| 32k/256 c=16 | median TPOT ms | 52.09 | 47.86 (-8.1%) |
| 1k/512 c=64 | decode tok/s | 2003 | 2130 (+6.3%) |
| 1k/512 c=64 | median TPOT ms | 30.00 | 27.56 (-8.1%) |
| 1k/512 c=256 | decode tok/s | 3697 | 4116 (+11.3%) |
| 1k/512 c=256 | median TPOT ms | 67.96 | 60.53 (-10.9%) |

### Accuracy (per build, same session)

Each rung of the ablation was also evaluated on its own: **gsm8k** (1319 questions, 5-shot, greedy, lm_eval `local-completions`

| build | gsm8k flexible-extract % | gsm8k strict-match % |
|---|---|---|
| main | 93.33 ± 0.69 | 93.10 ± 0.70 |
| +#55736 +#55737 | 93.25 ± 0.69 | 93.18 ± 0.69 |

## Duplicate-work check

`gh pr list --repo vllm-project/vllm --state open --search "GLM-5.3-Flash FlashKDA"` / `"glm5next KDA"`: no open PR wires FlashKDA for GLM-5.3-Flash (#55224 is the KDA conv weight refit fix).
- vllm-project/vllm#55364 (FlashInfer KDA kernels) is Kimi-K3 only (`vllm/models/kimi_k3/...`) and adds a `--kda-prefill-backend` engine arg; this PR touches only `vllm/models/glm5next/...`. Once #55364 lands, the `additional_config.kda_prefill_backend` knob here should move to that arg.

## Tests

- `bench/kda_numerics.py`: FlashKDA vs the naive recurrent reference (from `tests/models/kimi_k3/test_kda.py`) mean|Δ| 2e-5, max 4.8e-4 on outputs, same as the Triton chunk path vs the reference (T=48, 2 sequences, lower_bound -5). Note the Triton chunk path writes its output into `v`'s buffer, so the comparison clones `v`.
- `pytest tests/models/kimi_k3/test_kda.py -k flashkda`: 2 passed (extension sanity).
- e2e: chat sanity; prompt-logprob A/B vs main within ~1-2.5x the main-vs-main noise floor with identical next tokens; **gsm8k 1319 questions (5-shot, greedy): main 93.03% ± 0.70, full series 92.65% ± 0.72** (measured on the three PRs together).
- `pre-commit run ruff-check / ruff-format`: passed.

## Notes for review

- FlashKDA hardcodes dense q/k/v strides, so the three `.contiguous()` copies of the conv-output slices remain (~115 us/layer per 16k chunk); a follow-up could run the short conv per q/k/v into separate buffers.
- AI assistance: developed with Claude and reviewed by the submitter

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/models/glm5next/nvidia/kda.py [hunk 1]
```diff
@@ -533,26 +630,42 @@ def _rearr(x):
...
-                A_log=self.A_log,
-                g_bias=self.dt_bias,
-                initial_state=initial_state,
-                output_final_state=True,
-                use_qk_l2norm_in_kernel=True,
-                cu_seqlens=non_spec_query_start_loc,
-                safe_gate=safe_gate,
-                lower_bound=lower_bound,
-            )
+            if self.kda_prefill_backend == "flashkda" and not use_spec:
```
Comment (ZJY0516) [inline vllm/models/glm5next/nvidia/kda.py]: I remember flashkda also supports spec decoding?

Comment (JaredforReal) [reply vllm/models/glm5next/nvidia/kda.py]: I remember not,need more profiling and optimization on SpecDecode path tho..

Comment (ZJY0516) [reply vllm/models/glm5next/nvidia/kda.py]: you can check this: `vllm/models/kimi_k3/nvidia/kda.py`, no such guard

Comment (JaredforReal) [reply vllm/models/glm5next/nvidia/kda.py]: thanks, will work on it

**Comment by NNNtrance** [comment]
Field report from a different stack, in case it helps the review: three DGX Spark GB10 nodes (sm_121, arm64, CUDA 13.0) serving GLM-5.3-Flash as an EXL3 4.05 bpw checkpoint at TP=3 + expert parallelism, vLLM pinned at 487ecf187 with this PR's `kda.py` change ported as a startup patch. The `_flashkda_C` in our image is the 26 Aug build; its family-120 cubin runs on sm_121 without an sm_121 target.

- Kernel microbench at production shape (22 KDA heads per rank, head_dim 128, bf16, bounded gate), 2048 tokens: 2.47 ms on the Triton chunk path → 0.97 ms with FlashKDA, the three `.contiguous()` copies included (2.55×). Numerics against the Triton path: mean |Δ| 1.2e-5, max 9.8e-4 on the outputs, i.e. bf16-ULP scale.
- End to end, symmetric A/B (both arms booted the same way, only the backend differs; 24 fresh ~7,000-token prompts kept in flight, 60 s windows, zero prefix-cache hits): prefill 1,753.5 / 1,754.0 → 1,867.4 / 1,868.8 tok/s (+6.5 %), TTFT for a 7K prompt 4.19 → 3.98 s (−5.1 %), single-stream decode unchanged (61.2 vs 61.3 tok/s), KV pool unchanged, quality gates unchanged (correctness probe 10/10, code exam 12/12, 6-depth needle 6/6, vision OK).
- Our prefill profile puts the replaced kernels at about 6 % of GPU time, so we expected +3.5–4 %; the measured +6.5 % is more than that share explains and we have not root-caused the difference.
- One small suggestion: log which backend was resolved. We added a one-line `kda_prefill_backend=flashkda|triton` at startup; without it an A/B cannot be verified from the logs.

Full write-up, harness and the port: https://github.com/NNNtrance/GLM-5.3-Flash-EXL3-DGX-Spark/blob/main/results/gates/flashkda-ab-10sep.md. It is in our production as of today. Thanks for the port.

**Review by ZJY0516 — APPROVED** [review]

**Comment by gaby** [comment]
@JaredforReal I noticed you are using `--attention-backend FLASHINFER_MLA_SPARSE` but none of the official recipes do. Is that better for performance?

**Comment by JaredforReal** [comment]
@gaby nope, now glm-5.3-flash in vllm has only one attention backend, which is FLASHINFER_MLA_SPARSE
Adding the flag is only for clarity
